src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+        const result = new Texture(device, {
+            name: 'skyboxFaces',
+            cubemap: true,
+            width: size,
+            height: size,
+            type: TEXTURETYPE_RGBM,
+            addressU: ADDRESS_CLAMP_TO_EDGE,
+            addressV: ADDRESS_CLAMP_TO_EDGE,
+            fixCubemapSeams: fixCubemapSeams,
+            mipmaps: false
+        });
+
+        reprojectTexture(source, result, {
+            numSamples: 1024
+        });
+
+        return result;
+    }
+
+    /**
+     * @private
+     * @function
+     * @name generateLightingSource
+     * @description Create a texture in the format needed to precalculate lighting data.
+     * @param {Texture} source - The source texture. This is either a 2d texture in equirect format or a cubemap.
+     * @returns {Texture} The resulting cubemap.
+     */
+    static generateLightingSource(source) {
+        const device = source.device;
+        const format = lightingSourcePixelFormat(device);
+
+        const result = new Texture(device, {
+            name: 'lightingSource',
+            cubemap: true,
+            width: 128,
+            height: 128,
+            format: format,
+            type: format === PIXELFORMAT_R8_G8_B8_A8 ? TEXTURETYPE_RGBM : TEXTURETYPE_DEFAULT,
+            addressU: ADDRESS_CLAMP_TO_EDGE,
+            addressV: ADDRESS_CLAMP_TO_EDGE,
+            fixCubemapSeams: fixCubemapSeams,
+            mipmaps: false
+        });
+
+        // copy into top level
+        reprojectTexture(source, result, {
+            numSamples: source.mipmaps ? 1 : 1024
+        });
+
+        // generate mipmaps
+        generateMipmaps(result);
+
+        return result;
+    }
+
+    /**
+     * @private
+     * @function
+     * @name generateReflection
+     * @description Generate a prefiltered reflection texture for use in IBL.
+     * @param {Texture} source - The source lighting texture, as returned from generateLightingSource.
+     * @param {object} [options] - The optional options object.
+     * @param {number} [options.numSamples] - Number of samples. Default is 1024.
+     * @param {string} [options.distribution] - The specular distribution, either 'phong' or 'ggx'. Default is 'ggx'.
+     * @returns {Texture} The prefiltered cubemap
+     */
+    static generateReflection(source, options) {
+        const device = source.device;
+        const format = lightingPixelFormat(device);
+        const result = createCubemap(device, 128, format);
+
+        // generate the top level specular - just a copy of source
+        reprojectTexture(source, result, {
+            numSamples: 1
+        });
+
+        // generate convolved mipmaps
+        const levels = calcLevels(result.width, result.height);
+        for (let i = 1; i < levels; ++i) {
+            const level = createCubemap(device, Math.max(1, result.width >> i), format);
+            reprojectTexture(source, level, {
+                numSamples: options?.numSamples || 1024,
```

Comment:
should we default to less than 1024?

Reply:
Think 1024 is a good default. Users doing dynamically generated stuff might want to go lower and trade quality, but some HDR environments need 1024.

That said, we can definitely test and tweak as we go.